### PR TITLE
New check E0244: Extending inherited Enum class

### LINKED ISSUE #5501: `invalid-enum-extension` to detect subclassing an enum class
### Current problem

In general `Enum` classes cannot be subclassed. At runtime Python will raise a `TypeError`.

```py
from enum import Enum

class C(Enum):
    A = 1
    B = 2

class D(C):
    C = 3
```

```
TypeError: D: cannot extend enumeration 'C'
```

### Desired solution

Detect that an Enum class is subclassed and emit an error.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jpy-git**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Closes #5501. Prevents invalid extension of Enum classes.

### CODE REVIEW

File: doc/data/messages/i/invalid-enum-extension/good.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from enum import Enum
+
+
+class A(Enum):
+    x = 1
+    y = 2
+    z = 3
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/invalid-enum-extension/good.py:7]:
```suggestion


class A(Enum):
    x = 1
    y = 2


class B(Enum):
    x = 1
    y = 2
    z = 3
```
I suppose if you need two classes the solution is to duplicate ?

File: doc/data/messages/i/invalid-enum-extension/bad.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from enum import Enum
+
+
+class A(Enum):
+    x = 1
+    y = 2
+
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/invalid-enum-extension/bad.py:7]:
```suggestion


class A(Enum):
    x = 1
    y = 2


```
pep8

File: pylint/checkers/classes/class_checker.py [hunk 3]
```diff
@@ -835,6 +842,15 @@ def _check_proper_bases(self, node):
             ):
                 self.add_message("inherit-non-class", args=base.as_string(), node=node)
 
+            if (
+                isinstance(ancestor, nodes.ClassDef)
+                and ancestor.is_subtype_of("enum.Enum")
+                and any(isinstance(stmt, nodes.Assign) for stmt in ancestor.body)
+            ):
+                self.add_message(
+                    "invalid-enum-extension", args=ancestor.name, node=node
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:851]:
```suggestion
                    "invalid-enum-extension", args=ancestor.name, node=node, confidence=HIGH
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Looks pretty good already !

File: doc/data/messages/i/invalid-enum-extension/bad.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+class B(A):  # [invalid-enum-extension]
+    z = 3
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/invalid-enum-extension/bad.py:10]:
```suggestion
from enum import Enum


class Color(Enum):
    ORANGE = 1
    CHERRY = 2


class Fruit(Color):  # [invalid-enum-extension]
    APPLE = 3
```

File: doc/data/messages/i/invalid-enum-extension/good.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+class A(Enum):
+    x = 1
+    y = 2
+
+
+class B(Enum):
+    x = 1
+    y = 2
+    z = 3
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/i/invalid-enum-extension/good.py:12]:
```suggestion
from enum import Enum


class Color(Enum):
    ORANGE = 1
    CHERRY = 2


class Fruit(Enum):
    ORANGE = 1
    CHERRY = 2
    APPLE = 3
```

**Review by Pierre-Sassoulas — COMMENTED** [review]
Let's make the wrong example feel wrong 😄

File: doc/data/messages/i/invalid-enum-extension/bad.py (hunk 4 above)
Comment (jpy-git) [reply doc/data/messages/i/invalid-enum-extension/bad.py:10]: Haha yeah I like the more interesting naming!

File: pylint/checkers/classes/class_checker.py [hunk 6]
```diff
@@ -835,6 +842,18 @@ def _check_proper_bases(self, node):
...
+            if (
+                isinstance(ancestor, nodes.ClassDef)
+                and ancestor.is_subtype_of("enum.Enum")
+                and any(isinstance(stmt, nodes.Assign) for stmt in ancestor.body)
+            ):
+                self.add_message(
+                    "invalid-enum-extension",
+                    args=ancestor.name,
+                    node=node,
+                    confidence=HIGH,
```
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:854]: Just quickly checking in: this should be `INFERENCE` see L832.

**Review by Pierre-Sassoulas — APPROVED** [review]

File: doc/whatsnew/2.14.rst [hunk 7]
```diff
@@ -12,6 +12,10 @@ Summary -- Release highlights
 New checkers
 ============
 
+* ``invalid-enum-extension``: Used when class tries to extend an inherited Enum class.
```
Comment (DanielNoord) [inline doc/whatsnew/2.14.rst:15]:
```suggestion
* ``invalid-enum-extension``: Used when a class tries to extend an inherited Enum class.
```

File: ChangeLog [hunk 8]
```diff
@@ -10,6 +10,10 @@ Release date: TBA
   Put new features here and also in 'doc/whatsnew/2.14.rst'
 
 
+* ``invalid-enum-extension``: Used when class tries to extend an inherited Enum class.
```
Comment (DanielNoord) [inline ChangeLog:13]:
```suggestion
* ``invalid-enum-extension``: Used when a class tries to extend an inherited Enum class.
```

File: pylint/checkers/classes/class_checker.py [hunk 9]
```diff
@@ -626,6 +626,12 @@ def _has_same_layout_slots(slots, assigned_value):
         "Used when an invalid object is assigned to a __class__ property. "
         "Only a class is permitted.",
     ),
+    "E0244": (
+        'Extending inherited Enum class "%s"',
+        "invalid-enum-extension",
+        "Used when class tries to extend an inherited Enum class. "
```
Comment (DanielNoord) [inline pylint/checkers/classes/class_checker.py:632]:
```suggestion
        "Used when a class tries to extend an inherited Enum class. "
```

**Review by DanielNoord — APPROVED** [review]
Missing one word. Rest looks good!

Thanks @jpy-git!

**Pull request merged**